server/src/main/java/org/candlepin/controller/ModeManagerImpl.java

```diff
@@ -0,0 +1,92 @@
+/**
+ * Copyright (c) 2009 - 2012 Red Hat, Inc.
+ *
+ * This software is licensed to you under the GNU General Public License,
+ * version 2 (GPLv2). There is NO WARRANTY for this software, express or
+ * implied, including the implied warranties of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+ * along with this software; if not, see
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+ *
+ * Red Hat trademarks are not licensed under GPLv2. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package org.candlepin.controller;
+
+import com.google.inject.Singleton;
+import org.candlepin.common.exceptions.SuspendedException;
+import org.candlepin.model.CandlepinMode;
+import org.candlepin.model.CandlepinMode.Mode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+
+/**
+ * This class drives Suspend Mode functionality. Suspend Mode is a
+ * mode where Candlepin stops serving requests and only responds to
+ * Status resource.
+ *
+ * @author fnguyen
+ */
+@Singleton
+public class ModeManagerImpl implements ModeManager {
+    private static Logger log = LoggerFactory.getLogger(ModeManagerImpl.class);
+    private CandlepinMode cpMode = new CandlepinMode(new Date(), Mode.NORMAL, "Starting into NORMAL mode.");
+    private List<ModeChangeListener> listeners = new ArrayList<ModeChangeListener>();
+
+    @Override
+    public Mode getCurrentMode() {
+        return cpMode.getMode();
+    }
+
+    @Override
+    public CandlepinMode getCurrentCandlepinMode() {
+        return cpMode;
+    }
+
+    @Override
+    public boolean enterMode(Mode m, String reason) {
+        log.info("Entering new mode {} for reason {}", m, reason);
+
+        Mode previousMode = cpMode.getMode();
+
+        if (previousMode != m) {
+            fireModeChangeEvent(m);
+        }
+
+        cpMode.setChangeTime(new Date());
```

Comment:
ack, I made CandlepinMode class attributes final and removed setter methods and new CandlepinMode is created when changing modes now.